usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2690,6 +2698,285 @@ static CK_RV ccatok_pkey_add_attrs(STDLL_TokData_t * tokdata, TEMPLATE *template
+        TRACE_ERROR("Invalid/unknown cca token, cannot get key type\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Create the protected key by iterating over all APQNs */
+    memset(&pkey_wrap_handler_data, 0, sizeof(pkey_wrap_handler_data_t));
+    pkey_wrap_handler_data.tokdata = tokdata;
+    pkey_wrap_handler_data.secure_key = skey_attr->pValue;
+    pkey_wrap_handler_data.secure_key_len = skey_attr->ulValueLen;
+    pkey_wrap_handler_data.pkey_buf = (CK_BYTE *)&pkey_buf;
+    pkey_wrap_handler_data.pkey_buflen_p = &pkey_buflen;
+    pkey_wrap_handler_data.keytype = key_type;
+
+    ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
+                               &pkey_wrap_handler_data);
+
+    if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+        TRACE_ERROR("cca_iterate_adapters failed or no APQN could create the pkey.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Build new attribute for protected key */
+    ret = build_attribute(CKA_IBM_OPAQUE_PKEY, pkey_buf, pkey_buflen, &tmp_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("build_attribute failed with rc=0x%lx\n", ret);
+        ret = CKR_FUNCTION_FAILED;;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    *pkey_attr = tmp_attr;
+
+    return ret;
+}
+
+/*
+ * Save the current firmware wrapping key verification pattern in the tokdata:
+ * create a dummy test key, transform it into a protected key, and store the wk
+ * verification pattern in the tokdata.
+ */
+static CK_RV ccatok_pkey_get_firmware_wkvp(STDLL_TokData_t *tokdata)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL, *sec_attr = NULL;
+    CK_RV ret;
+    long return_code = 0, reason_code = 0;
+    long exit_data_len = 0, clear_key_bit_length = 0;
+    unsigned char exit_data[4];
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE];
+    long rule_array_count;
+    unsigned char *clear_key_value = NULL;
+    long key_name_length = 0;
+    unsigned char key_name[CCA_KEY_ID_SIZE] = { 0, };
+    long user_data_length = 0;
+    unsigned char user_data[64] = { 0, };
+    long token_data_length = 0;
+    long verb_data_length = 0;
+    unsigned char verb_data[64] = { 0, };
+    unsigned char the_key_token[CCA_KEY_TOKEN_SIZE] = { 0, };
+    long the_key_token_length = sizeof(the_key_token);
+    unsigned char key_type_1[8] = { ' ', };
+    unsigned char key_type_2[8] = { ' ', };
+    long key_name_1_length = 0;
+    long key_name_2_length = 0;
+    long user_data_1_length = 0;
+    long user_data_2_length = 0;
+    long kek_identifier_1_length = 0;
+    long kek_identifier_2_length = 0;
+    long the_2nd_key_token_length = 0;
+
+    /* Check CCA host library version: XPRTCPAC requires min 7.0.0 */
+    if (cca_data->version.ver < 7) {
+        TRACE_ERROR("CCA host lib is %d.%d.%d, but pkey support requires min 7.0.0\n",
```

Comment:
Maybe also issue a SYSLOG message here? 
If it fails because of that, then below code issues syslog message "Warning: Could not get mk_vp, protected key support not available.", but it does not show a reason. So if this message is issued to SYSLOG then the reason would be know. It might not be too seldom that uses run with a CCA host library earlier than version 7.....

Reply:
Done.